cli/azd/cmd/init.go

```diff
@@ -104,7 +104,7 @@ type initAction struct {
 	lazyEnvManager  *lazy.Lazy[environment.Manager]
 	console         input.Console
 	cmdRun          exec.CommandRunner
-	gitCli          git.GitCli
+	gitCli          *git.Cli
```

Comment:
Do we have a pointer to ensure a singleton tool is used/injected everywhere?

Reply:
That was not the reason I used a pointer here (and, to note, that before the underlying value was logically `*git.Cli`, because the pointer type was the type that implemented the `git.GitCli` interface (since the receiver of each func was a pointer)).  I used a pointer here because that's the general go idiom for types like this - It's what the `New` function returns - and mimics other types like `*http.Client`.

If we typed this as `git.Cli` instead, nothing would fundamentally change - the type itself is stateless, the only "state" on it is the command runner - which is assigned at construction and never changes, but it still felt like `*git.Cli` would be preferred to `git.Cli`.

At some level, this type is really just about behaviors, not data, and so I really don't think about the difference between reference and value semantics.

Reply:
thanks for the explanation :)

Yeah, the before was logically a pointer, but that was mandated by the interface itself.

When moving to concrete implementation, I naturally wonder about pointer v/s non-pointer.

When you say `that's the general go idiom for types like this`, it makes me wonder...  My understanding is that go has a preference on copying values, instead of copying references, as the copies are generated really fast. So, makes me wonder.

Some tools are not 100% stateless, like `bicep` or `gh`, as they would find the path for the binary or download it when missing.  Having a singleton for those tools seems like the best approach. We want to control only one instance is created and used along azd execution.

But, the thing that keeps me wondering is... are we being careless about thread safety?  
I understand we are not currently using any multi-threading for any command, but should we be thinking on scenarios where we should prefer having the IoC container to inject only component copies and not reference copies?

I can't help to think that all those functions taking references should be checking for nil before using the reference. The fact that injection is controlled from a single place with IoC helps me to feel a little better, but there is still the feeling there... Because, in my mind, when a function takes a reference, it should be because it has one functionality for when the reference is null and one for when it is not null. It should not be because it is idiomatic to take a reference and just ignore handling the nil case.